### PR TITLE
Consume IDPs in UDO

### PR DESCRIPTION
**Pull request by johnmcollier**
Signed-off-by: John Collier <John.J.Collier@ibm.com>

## What is the purpose of this change? What does it change?

This PR updates UDO to pull down the Iterative-Dev Pack for the selected component and consume it when deploying applications.

## Was the change discussed in an issue?
fixes #???
N/A

## How to test changes?
Test results:
```
Johns-MacBook-Pro-3:kdo johncollier$ make test
go test -race github.com/redhat-developer/odo-fork/cmd/udo github.com/redhat-developer/odo-fork/pkg/application github.com/redhat-developer/odo-fork/pkg/application/labels github.com/redhat-developer/odo-fork/pkg/build github.com/redhat-developer/odo-fork/pkg/catalog github.com/redhat-developer/odo-fork/pkg/common github.com/redhat-developer/odo-fork/pkg/component github.com/redhat-developer/odo-fork/pkg/component/labels github.com/redhat-developer/odo-fork/pkg/config github.com/redhat-developer/odo-fork/pkg/idp github.com/redhat-developer/odo-fork/pkg/kclient github.com/redhat-developer/odo-fork/pkg/kdo/cli github.com/redhat-developer/odo-fork/pkg/kdo/cli/catalog github.com/redhat-developer/odo-fork/pkg/kdo/cli/catalog/list github.com/redhat-developer/odo-fork/pkg/kdo/cli/catalog/search github.com/redhat-developer/odo-fork/pkg/kdo/cli/component github.com/redhat-developer/odo-fork/pkg/kdo/cli/component/ui github.com/redhat-developer/odo-fork/pkg/kdo/cli/config github.com/redhat-developer/odo-fork/pkg/kdo/cli/project github.com/redhat-developer/odo-fork/pkg/kdo/cli/ui github.com/redhat-developer/odo-fork/pkg/kdo/cli/version github.com/redhat-developer/odo-fork/pkg/kdo/genericclioptions github.com/redhat-developer/odo-fork/pkg/kdo/util github.com/redhat-developer/odo-fork/pkg/kdo/util/validation github.com/redhat-developer/odo-fork/pkg/log github.com/redhat-developer/odo-fork/pkg/log/fidget github.com/redhat-developer/odo-fork/pkg/preference github.com/redhat-developer/odo-fork/pkg/project github.com/redhat-developer/odo-fork/pkg/storage github.com/redhat-developer/odo-fork/pkg/storage/labels github.com/redhat-developer/odo-fork/pkg/testingutil github.com/redhat-developer/odo-fork/pkg/util
?   	github.com/redhat-developer/odo-fork/cmd/udo	[no test files]
ok  	github.com/redhat-developer/odo-fork/pkg/application	(cached)
ok  	github.com/redhat-developer/odo-fork/pkg/application/labels	(cached)
?   	github.com/redhat-developer/odo-fork/pkg/build	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/catalog	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/common	[no test files]
ok  	github.com/redhat-developer/odo-fork/pkg/component	4.244s
ok  	github.com/redhat-developer/odo-fork/pkg/component/labels	(cached)
ok  	github.com/redhat-developer/odo-fork/pkg/config	1.229s
?   	github.com/redhat-developer/odo-fork/pkg/idp	[no test files]
ok  	github.com/redhat-developer/odo-fork/pkg/kclient	(cached)
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/catalog	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/catalog/list	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/catalog/search	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/component	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/component/ui	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/config	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/project	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/ui	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/cli/version	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/kdo/genericclioptions	[no test files]
ok  	github.com/redhat-developer/odo-fork/pkg/kdo/util	(cached)
ok  	github.com/redhat-developer/odo-fork/pkg/kdo/util/validation	1.152s
?   	github.com/redhat-developer/odo-fork/pkg/log	[no test files]
?   	github.com/redhat-developer/odo-fork/pkg/log/fidget	[no test files]
ok  	github.com/redhat-developer/odo-fork/pkg/preference	1.142s
ok  	github.com/redhat-developer/odo-fork/pkg/project	(cached)
ok  	github.com/redhat-developer/odo-fork/pkg/storage	(cached)
ok  	github.com/redhat-developer/odo-fork/pkg/storage/labels	(cached)
ok  	github.com/redhat-developer/odo-fork/pkg/testingutil	(cached)
ok  	github.com/redhat-developer/odo-fork/pkg/util	(cached)
```
1. `make install`
2. `udo create nodejs`
    - Verify that a `.udo` folder exists and contains both a config.yaml and an idp.yaml
    - Verify that the idp.yaml is the correct one
3. `udo push`
    - Verify that the runtime image specified in the idp.yaml is used in the runtime deployment.

### CODE REVIEW

File: pkg/idp/types.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+// TypeDetection defines the rules used to determine if a project is compatible with a specific IDP
+type TypeDetection struct {
+	PathSelector       []PathSelector `yaml:"pathSelector"`
+	TextStringsToMatch []string       `yaml:"textStringsToMatch"`
+}
+
+type PathSelector struct {
+	RootPath         string `yaml:"rootPath"`
+	FileNameWildcard string `yaml:"filenameWildcard"`
```
Comment (jgwest) [inline pkg/idp/types.go]: Should perhaps be `FilenameWildcard` rather than `FileNameWildcard`

File: pkg/idp/types.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	InternalPerformancePort string `yaml:"internalPerformancePort"`
+}
+
+type Logs struct {
+	Type string `yaml:"type"`
+	Path string `yaml:"path"`
+}
+
+type SharedTasks struct {
+	DisposeOfSharedContainersOnTaskComplete string `yaml:"disposeOfSharedContainersOnTaskComplete"`
```
Comment (jgwest) [inline pkg/idp/types.go]: `string` rather than `bool`?

File: pkg/idp/types.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+// UploadFilter specifies any file paths to be ignored when syncing files (like target/)
+type UploadFilter struct {
+	IgnoredPaths []string `yaml:"ignoredPaths"`
+}
+
+// TypeDetection defines the rules used to determine if a project is compatible with a specific IDP
+type TypeDetection struct {
+	PathSelector       []PathSelector `yaml:"pathSelector"`
+	TextStringsToMatch []string       `yaml:"textStringsToMatch"`
+}
```
Comment (jgwest) [inline pkg/idp/types.go]: Go representation differs here from the YAML schema (if you think the YAML schema should work differently, let me know..) 

A single `PathSelector` has multiple `TextStringsToMatch`, here is how the JSON looks:
```json
      "typeDetection": [
        {
          "pathSelector": {
            "rootPath": null,
            "filenameWildcard": null
          },
          "textStringsToMatch": [
            "net.wasdev.wlp.maven.plugins"
          ]
        }
      ]
```

So you would need a Go struct like:
```go
type TypeDetectionEntry struct {
	PathSelector        PathSelector   `yaml:"pathSelector"`
	TextStringsToMatch []string       `yaml:"textStringsToMatch"`
}
```

File: pkg/idp/types.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	BuildImage       string             `yaml:"buildImage"`
+	Command          string             `yaml:"command"`
+	WorkingDirectory string             `yaml:"workingDirectory"`
+	Logs             []Logs             `yaml:"logs"`
+	VolumeMappings   []Mappings         `yaml:"volumeMappings"`
+	RepoMappings     []Mappings         `yaml:"repoMappings"`
+	SourceMappings   []Mappings         `yaml:"sourceMappings"`
+	RunAsUser        int                `yaml:"runAsUser"`
+	Kubernetes       KubernetesSettings `yaml:"kubernetes"`
+}
```
Comment (jgwest) [inline pkg/idp/types.go]: Appears to be missing Env, ala:
```go
Env []EnvVars
```

File: pkg/idp/types.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	Name   string `yaml:"name"`
+	Labels string `yaml:"labels"`
+	size   int    `yaml:"size"`
+}
+
+// EnvVars represents a key/value mapping of environment vars to use in runtime and build containers
+type EnvVars struct {
+	key   string `yaml:"key"`
+	value string `yaml:"value"`
+}
```
Comment (jgwest) [inline pkg/idp/types.go]: Since this is a single key/value pair, it should be `EnvVar`

File: pkg/idp/types.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	Tasks     []SpecTask     `yaml:"tasks"`
+	Scenarios []SpecScenario `yaml:"scenario"`
+}
+
+// SpecDev ...
+type SpecDev struct {
+	Watched       WatchedFiles  `yaml:"watched"`
+	UploadFilter  UploadFilter  `yaml:"uploadFilter"`
+	TypeDetection TypeDetection `yaml:"typeDetection"`
+}
```
Comment (jgwest) [inline pkg/idp/types.go]: In the schema, `TypeDetection` is an array. Here is the corresponding JSON:
```json

      "typeDetection": [
        {
          "pathSelector": {
            "rootPath": null,
            "filenameWildcard": null
          },
          "textStringsToMatch": [
            "net.wasdev.wlp.maven.plugins"
          ]
        }
      ]
    },
```

File: go.mod [hunk 7]
```diff
@@ -21,6 +21,7 @@ require (
 	golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
 	gopkg.in/AlecAivazis/survey.v1 v1.8.5
 	gopkg.in/yaml.v2 v2.2.2
+	gopkg.in/yaml.v3 v3.0.0-20190905181640-827449938966
```
Comment (rajivnathan) [inline go.mod]: Can the old version be removed?

File: pkg/config/config.go [hunk 8]
```diff
@@ -101,20 +101,29 @@ type LocalConfigInfo struct {
 	configFileExists bool
 }
 
-func getLocalConfigFile(cfgDir string) (string, error) {
-	if env, ok := os.LookupEnv(localConfigEnvName); ok {
-		return env, nil
-	}
-
+// GetUDOFolder retrieves the absolute path of the .udo folder in the current directory udo is used in
+func GetUDOFolder(cfgDir string) (string, error) {
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply pkg/config/config.go]: Yeah, good idea.

File: pkg/idp/idp.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+)
+
+// Get loads in the project's idp.yaml from disk
+func Get() (*IDP, error) {
+	// Retrieve the IDP.yaml file
+	udoDir, err := config.GetUDOFolder("")
+	if err != nil {
+		return nil, fmt.Errorf("unabled to find .udo folder in current directory")
+	}
+	idpFile := path.Join(udoDir, "idp.yaml")
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply pkg/idp/idp.go]: Haha, good catch

File: pkg/idp/idp.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+		}
+	}
+
+	err = yaml.Unmarshal(idpBytes, &idp)
+	return &idp, err
+}
+
+// DownloadIDPYaml downloads the idp.yaml for the iterative-dev pack at the given URL
+func DownloadIDPYaml(idpURL string) error {
+	// Download the IDP index.json
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply pkg/idp/idp.go]: What about verifying we can unmarshall it to a struct?

File: go.mod (hunk 7 above)
Comment (johnmcollier) [reply go.mod]: Hmm, odd that the old version wasn't removed by go mod.

File: pkg/idp/idp.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (rajivnathan) [reply pkg/idp/idp.go]: Discussed this offline, that works. Will handle proper validation in a separate issue.

**Pull request merged**